[vNext] Redesing to material-ui 2.0

dmtrKovalenko opened the pull request:
<!-- Thanks so much for your time taking to contribute, your work is appreciated! ❤️ -->

This PR is a part of v4 development (#1293)

## Description

- [x] Common design changes like toolbar, dimensions, paddings, etc...
- [x] Redesign datepicker with toolbar
- [x] Change moving views logic to be inside the calendar and change year with triangle icon
- [x] Fix tests <-- In progress right now, moving our tests from enzyme to cypress components tests
- [x] Redesign timepicker
- [x] Redesign datetime picker 

Note that new layout for mobile/desktop variant will be in the next PR, here just changing everything inside the components.

> This PR hardly in development right now, but please @oliviertassinari take a brief look of the deployment preview for how it looks and feels

dmtrKovalenko committed:
Redesign datepicker toolbar

dmtrKovalenko committed:
Add optinal chaining support

dmtrKovalenko committed:
Redesign calendar view and switching between to year view

dmtrKovalenko committed:
Small bugfixes and improvements

dmtrKovalenko committed:
Add weekdays

dmtrKovalenko committed:
Redesign year selection

dmtrKovalenko committed:
Cleanup and remove not used files

dmtrKovalenko committed:
Add typings for style overrdies  of MuiBasePickerO, closes #1414

dmtrKovalenko committed:
New animations 😍

oliviertassinari commented on lib/src/DatePicker/DatePickerToolbar.tsx:
```diff
@@ -1,23 +1,28 @@
+      justifyContent: 'space-between',
+      paddingTop: 16,
+      paddingBottom: 16,
+      padding: props => (props.isLandscape ? 16 : undefined),
```
I would avoid using style function in any of the components, it's not ready for prime-time.

dmtrKovalenko replied:
:( I was happy that it will be possible to get rid of clsx 😢. Do you know when it will be ok? Because docs are saying that everything is stable enough.

oliviertassinari replied:
I wish we were able to use them in the core components too. We don't as we are speaking. I think that it would either require the styled-components migration or to use react-jss internally. Performance is OK but not great. It's fine for the Box, but I wouldn't use them as a systemic solution.

oliviertassinari commented on lib/src/DatePicker/DatePickerToolbar.tsx:
```diff
@@ -26,50 +31,51 @@ export const useStyles = makeStyles(
+      <Typography
+        color="inherit"
+        variant="overline"
+        children={title}
```
A funny coding style, was it on purpose 🙃?

dmtrKovalenko replied:
It allows to get rid of unnecessary spaces when passing down children props.

oliviertassinari replied:
So it's to optimize for readability when developing?

dmtrKovalenko replied:
Both. But actually this 
```jsx
<div> {children} </div>
```

Will render 2 spaces before and after the children. And also our behaviour looks more like spreading things down instead of rendering.

oliviertassinari commented on lib/src/_shared/icons/PenIcon.tsx:
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
+import SvgIcon, { SvgIconProps } from '@material-ui/core/SvgIcon';
+
+export const PenIcon: React.FC<SvgIconProps> = props => {
```
You might want to look at the `createSvgIcon()` approach we have. It can be useful. For instance, to forward ref.

dmtrKovalenko replied:
I do think about using @material-ui/icons again. I removed that dependency because there were a lot of problems with esm build. 
But AFAI everything should be ok right now? cc @TrySound

TrySound replied:
@oliviertassinari What is the approach you are talking about?

TrySound replied:
Got it. It's about avoiding boilerplate. Looks good to me.
https://github.com/mui-org/material-ui/blob/89687f38cae750650555772ba4d821c9084d8dfc/packages/material-ui/src/internal/svg-icons/ArrowDropDown.js#L7

oliviertassinari replied:
I don't think that the picker should be importing from @material-ui/icons either. It's too big of a dependency, especially for people that would want to customize the picker. We don't use the icons package either in the @material-ui/core package.

oliviertassinari commented on lib/src/views/Calendar/Calendar.tsx:
```diff
@@ -356,9 +71,152 @@ export const styles = (theme: Theme) => ({
+    [handleDaySelect, shouldDisableDate]
+  );
+
+  useKeyDown(Boolean(allowKeyboardControl && variant !== 'static'), {
```
There is some issues with this useKeyDown helper:
- It doesn't use the synthetic event, so he can't use `event.key` (React polyfills it)
- It doesn't listen to the active frame, I believe it wouldn't work when rendered in an iframe, popup

oliviertassinari commented on lib/src/views/Calendar/CalendarHeader.tsx:
```diff
@@ -1,124 +1,178 @@
-      marginTop: theme.spacing(0.5),
+      marginTop: theme.spacing(2),
       marginBottom: theme.spacing(1),
+      paddingLeft: theme.spacing(3),
```
I would suggest not to use the spacing helper in these components unless if somebody customizes it, it has an expected outcome.

dmtrKovalenko replied:
@oliviertassinari it is a nice point. I am really not clear about what spacing means. But we have some issues like #1141. How do you think should we support them or just ignore?

oliviertassinari replied:
Yes, #1128 is exactly why I would discourage using the spacing helper unless a different spacing has a clear expected outcome (we have very few cases on this on the main repository). I can give you one example `<Grid spacing={1} />`. I doubt there are many cases like this in the date picker.

oliviertassinari commented on lib/src/views/Calendar/CalendarHeader.tsx:
```diff
@@ -127,15 +181,11 @@ CalendarHeader.displayName = 'CalendarHeader';
-  disableNextMonth: PropTypes.bool,
 };
 
 CalendarHeader.defaultProps = {
```
Have you considered to use the prop spread default values instead of default props?

dmtrKovalenko replied:
It is primarily used for typescript typings. This is just so old code, will move it. (But there is a problem with auto generating default props values)

oliviertassinari replied:
We might have a solution for this. The default props generation for API purpose was solve in the mui-org/material-ui repository. @eps1lon took care of it.
It's a good reminder that we need to look into how we can extract the API from TypeScript sources, and not only JavaScript ones.

dmtrKovalenko replied:
Right now I created custom script that reads jsdoc tags on props like `@default "someValue"` and primarily use it and getting rid from default props.

eps1lon replied:
Our handler for `react-docgen` and default props via default values in object rest spread can be found in https://github.com/mui-org/material-ui/blob/6f7318cdd5c1c2887860c0d56212a3e42d107082/docs/src/modules/utils/defaultPropsHandler.js#L8

dmtrKovalenko replied:
@eps1lon don't you use `react-docgen-typescript`?

oliviertassinari commented on lib/src/views/Calendar/CalendarView.tsx:
```diff
@@ -0,0 +1,198 @@
+  minDate: unparsedMinDate,
+  maxDate: unparsedMaxDate,
+  loadingIndicator = <CircularProgress />,
+  ...rest
```
For consistency with the core?

```suggestion
  ...other
```

dmtrKovalenko committed:
Start cypress component tests

dmtrKovalenko committed:
Update lib/src/views/Calendar/CalendarView.tsx

Co-Authored-By: Olivier Tassinari <olivier.tassinari@gmail.com>

dmtrKovalenko committed:
Address PR feedback

dmtrKovalenko committed:
Add alternative way to display ampm controls

dmtrKovalenko committed:
Redesign TimePickers

dmtrKovalenko committed:
Redesign DateTimePicker

dmtrKovalenko committed:
Fix build errors

dmtrKovalenko committed:
Convert ModalDialog to hook based styles

dmtrKovalenko committed:
Remove unused stuff from dependencies

dmtrKovalenko committed:
Fix ts build errros

dmtrKovalenko committed:
Turn off empty interface rule

dmtrKovalenko committed:
Fix rollup error

dmtrKovalenko committed:
Fix displaying toolbars in firefox and safari

dmtrKovalenko committed:
Replace theme.spacing( ) with static values

dmtrKovalenko committed:
Fix examples with custom theme

dmtrKovalenko committed:
Fix a bunch of tests

dmtrKovalenko committed:
Fix tests

dmtrKovalenko committed:
Merge branch 'next' into feature/redesign-md2

dmtrKovalenko commented:
@oliviertassinari could you take a look to deployment preview and visual changes? There are quite a lot of code changes :)

dmtrKovalenko committed:
Rename cypress lib test file

dmtrKovalenko committed:
Fix CI

dmtrKovalenko committed:
Fix ts build errors

oliviertassinari commented:
@dmtrKovalenko Yes, I'm having a look! Well, I will try 🙃 

![Capture d’écran 2019-12-20 à 00 50 13](https://user-images.githubusercontent.com/3165635/71218719-dde51880-22c2-11ea-8eb2-18e9d38152d8.png)

oliviertassinari commented on docs/pages/demo/datepicker/ServerRequest.example.jsx:
```diff
@@ -26,7 +26,7 @@ function ServerRequest() {
         label="With server data"
         value={selectedDate}
-        onChange={handleDateChange}
+        onChange={date => handleDateChange(date)}
```
I'm wondering if we shouldn't make it clear that we don't expect a returned value.

```suggestion
        onChange={date => {
          handleDateChange(date)
        }}
```

dmtrKovalenko replied:
I dont think it will help somebody. I think 99% of js developers will be confused by such examples

oliviertassinari commented on docs/pages/guides/CssOverrides.example.jsx:
```diff
@@ -2,8 +2,7 @@
 import { DatePicker } from '@material-ui/pickers';
-import { createMuiTheme } from '@material-ui/core';
-import { ThemeProvider } from '@material-ui/styles';
+import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles';
```
We have removed `MuiThemeProvider` from the documentation.

```suggestion
import { createMuiTheme, ThemeProvider } from '@material-ui/core/styles';
```

dmtrKovalenko replied:
oh god :)

oliviertassinari commented on docs/pages/guides/CssThemeSpacing.example.jsx:
```diff
@@ -1,19 +1,18 @@
 
-const defaultMaterialTheme = createMuiTheme({
+const muiTheme = createMuiTheme({
   spacing: 2,
```
Should this be a visual regression test rather than a demo? But I'm not sure a visual test is needed. On the main repository, we had very few bug report / concern on this end. In https://material-ui.com/customization/density/#explore-theme-density, people can tweak the density.

dmtrKovalenko replied:
Actually there were some issues with it in pickers #1128. So this is more like a proof that we are not completely broken with custom spacing. But I agree that it needs to be a test rather than a demo

oliviertassinari commented on lib/src/DatePicker/DatePickerToolbar.tsx:
```diff
@@ -26,50 +23,32 @@ export const useStyles = makeStyles(
-  setOpenView,
   isLandscape,
-  openView,
+  title = 'SELECT DATE',
```
This would make a good new addition to https://material-ui.com/guides/localization/.
What about using lower case, so that people can more easily tweak the capitalization with CSS. For instance, they might have globally disabled the uppercase of the button.

```suggestion
  title = 'Select date',
```

oliviertassinari commented on lib/src/DateTimePicker/DateTimePicker.tsx:
```diff
@@ -14,9 +14,9 @@ import {
 } from '../Picker/makePickerWithState';
 
-export type DateTimePickerView = 'year' | 'date' | 'month' | 'hours' | 'minutes';
+export type DateTimePickerView = 'year' | 'date' | 'month' | 'hours' | 'minutes' | 'seconds';
```
😍

oliviertassinari commented on lib/src/DateTimePicker/DateTimePickerToolbar.tsx:
```diff
@@ -33,81 +45,55 @@ export const DateTimePickerToolbar: React.FC<ToolbarComponentProps> = ({
   dateRangeIcon,
   timeIcon,
-  onChange,
+  title = 'SELECT DATE & TIME',
```
```suggestion
  title = 'Select date & time',
```

I'm wondering if we should rename the prop to `titleText` to something that starts to look like a convention:
- backIconButtonText
- labelRowsPerPage
- labelDisplayedRows
- nextIconButtonText
- getLabelText
- clearText
- closeText
- loadingText
- noOptionsText
- openText

oliviertassinari replied:
We also need to be careful when using DOM attribute names.

dmtrKovalenko replied:
I thought this stuff should be filled from `label` coming to input. And then be overridable with custom prop

oliviertassinari commented on lib/src/TimePicker/TimePickerToolbar.tsx:
```diff
@@ -139,33 +133,34 @@ export const TimePickerToolbar: React.FC<ToolbarComponentProps> = ({
           })}
         >
           <ToolbarButton
+            data-test-id="toolbar-am-btn"
```
For reference, we have a `data-mui-test` convention in the main repository to solve this problem (+ a babel plugin to remove it before releasing). It would be great to unify the approach.

@eps1lon should we rename `data-mui-test` to `data-testid`, to match @testing-library default behavior?
```suggestion
            data-testid="toolbar-am-btn"
```

dmtrKovalenko replied:
Actually I don't like @testing-library so now working on improving cypress with component tests https://github.com/cypress-io/cypress/pull/5923. 
Going to try use it with pickers.
> P.S `data-testid` is incorrect word for spell checkers, but I don't care about it

oliviertassinari replied:
So, what do you think that we should do here? `data-testid` seems to match https://docs.cypress.io/guides/references/best-practices.html.

dmtrKovalenko replied:
Hehe, my bad :DDDD will rename it

oliviertassinari commented on lib/src/_shared/ModalDialog.tsx:
```diff
@@ -38,76 +63,56 @@ export const ModalDialog: React.SFC<ModalDialogProps & WithStyles<typeof styles>
+  const classes = useStyles();
+  return (
+    <Dialog
+      role="dialog"
```
Is the role needed?

dmtrKovalenko replied:
Haha, this line was added more than 2 years ago, sure it is not needed

oliviertassinari commented on lib/src/_shared/ModalDialog.tsx:
```diff
@@ -38,76 +63,56 @@ export const ModalDialog: React.SFC<ModalDialogProps & WithStyles<typeof styles>
-      )}
+      <DialogActions
+        classes={{
+          root: clsx({
```
Prefer className when possible :)

oliviertassinari commented on lib/src/wrappers/ModalWrapper.tsx:
```diff
@@ -65,7 +65,7 @@ export const ModalWrapper: React.FC<ModalWrapperProps<any>> = ({
   ...other
 }) => {
   useKeyDown(open, {
```
I'm wondering. Why not using a onKeydown listener on the element?

oliviertassinari commented on lib/src/views/Clock/Clock.tsx:
```diff
@@ -167,8 +60,163 @@ export const styles = (theme: Theme) =>
+      typeof e.buttons === 'undefined' ? e.nativeEvent.which === 1 : e.buttons === 1;
+
+    if (isButtonPressed) {
+      setTime(e.nativeEvent, false);
```
It's the first time I see the usage of `nativeEvent`. Do you have a specific use case over using `event`?

dmtrKovalenko replied:
Yes, react is not including some props from event like `changedTouches` 
```
offsetX = e.changedTouches[0].clientX - rect.left;
offsetY = e.changedTouches[0].clientY - rect.top;
```

same for `which`, which (🤣) is working only in safari

oliviertassinari replied:
I don't have the whole context, it might be something to dig into. It doesn't seem to be an issue with the `Slider` component that uses `changedTouches` too.

dmtrKovalenko replied:
Just checked with pickers, it doesn't works. BTW we are using also 
```
const rect = e.target.getBoundingClientRect();
```
which is more likely to be a problem. I don't remember why it was added. But it is required (it is crashing with event)

oliviertassinari replied:
Should it be `event.currentTarget.getBoundingClientRect()` here?

dmtrKovalenko replied:
![image](https://user-images.githubusercontent.com/16926049/71250697-f5ef8300-2328-11ea-8140-413d1a1bb9d9.png)

`e.changedTouches` is undefined for synthetic event

oliviertassinari replied:
For a mouse event, this sounds correct (mouse up). Should we use clientX and clientY for mouse events?

dmtrKovalenko replied:
Yes, because we need to get mouse position relative to the container for dragging. `offsetX` and `offsetY` will not work for all browsers

oliviertassinari commented on lib/src/views/Clock/Clock.tsx:
```diff
@@ -167,8 +60,163 @@ export const styles = (theme: Theme) =>
+    e.stopPropagation();
+    // MouseEvent.which is deprecated, but MouseEvent.buttons is not supported in Safari
+    const isButtonPressed =
+      typeof e.buttons === 'undefined' ? e.nativeEvent.which === 1 : e.buttons === 1;
```
I believe `event.button === 0` could work cross platform.

dmtrKovalenko replied:
We need e.buttons === 1

oliviertassinari replied:
I'm wondering because react-router checks `event.button` to only consider the main click (0).

oliviertassinari commented on lib/src/views/Clock/Clock.tsx:
```diff
@@ -167,8 +60,163 @@ export const styles = (theme: Theme) =>
     },
-  });
+    amButton: {
+      zIndex: 24,
```
Does 24 answer to a specific logic? Do you need the property in the first place?

dmtrKovalenko replied:
Thanks its debugging stuff

oliviertassinari commented on lib/src/_shared/ModalDialog.tsx:
```diff
@@ -38,76 +63,56 @@ export const ModalDialog: React.SFC<ModalDialogProps & WithStyles<typeof styles>
-    </DialogActions>
-  </Dialog>
-);
+        {cancelLabel && (
```
I'm wondering about a new name for the localization wordings (for consistency, as detailed above).

```suggestion
        {cancelText && (
```

dmtrKovalenko replied:
Let's pull this stuff out to separate PR with `localization support`. This one is so giant. I have a lot of thoughts about it, like providing localized formats for dates out of the box with localization. It will be a killer feature

oliviertassinari replied:
Yes, I agree!

oliviertassinari commented on lib/src/_shared/ToolbarButton.tsx:
```diff
@@ -1,24 +1,31 @@
 }
 
+export const useStyles = makeStyles(
+  _ => ({
```
No theme?

```suggestion
  {
```

dmtrKovalenko replied:
Agree. 
But don't like such approach. Let me explain. From the perspective of changing styles from using theme or not. To change it you need 1) remove function 2) remove the `)` from the end of expression. 
I would like to use only 1 way to create theme. 
> Also for reference from the algebraic type system perspective we should avoid such polymorphic functions, cause they are giving more friction.

oliviertassinari replied:
Remembers me when we were switching between class components and functions components, eww, it so much better with hooks now.

oliviertassinari commented on lib/src/views/Calendar/DayWrapper.tsx:
```diff
@@ -23,10 +23,9 @@ const DayWrapper: React.FC<DayWrapperProps> = ({
       role="presentation"
       onClick={dayInCurrentMonth && !disabled ? handleClick : undefined}
       onKeyPress={dayInCurrentMonth && !disabled ? handleClick : undefined}
+      children={children}
```
For the convention in the React community and for people reading the source, I think that the previous approach would be less destabilizing.

oliviertassinari commented on lib/src/views/Calendar/FadeTransitionGroup.tsx:
```diff
@@ -0,0 +1,76 @@
+  children: React.ReactChild;
+}
+
+const animationDuration = 500;
```
Should we use a transition duration from on the of the specification (theme.transitions.duration.x)?

dmtrKovalenko replied:
```
shortest: 150
shorter: 200
short: 250
standard: 300
complex: 375
enteringScreen: 225
leavingScreen: 195
```

All these durations are too fast for this animation.

oliviertassinari reviewed (commented):
This is a first batch of review, I have only looked at the source code. I cover some generic questions, and details.

Tomorrow, I will take more time to look at the product DX and UX, I think that I will be able to do more important feedback with this second phase :).

Well done so far 👌

dmtrKovalenko committed:
Replace global keydown handler in favor of per component

dmtrKovalenko committed:
Fix timepicker switching to next view on scroll

dmtrKovalenko committed:
Use object to create styles without theme

dmtrKovalenko committed:
Rename MuiThemeProvider to ThemeProvider

dmtrKovalenko committed:
Remove zIndex from am/pm buttons

dmtrKovalenko committed:
Fix ts error

dmtrKovalenko committed:
Fix ts issues

dmtrKovalenko committed:
[CI] Split build and lint to different workflows

dmtrKovalenko committed:
Attach workspace for new jobs

dmtrKovalenko committed:
Remove console.log

oliviertassinari reviewed (commented):
A second path, only looking at the end result. I assume the objective is to match the latest specification regarding the experience on mobile. I have noticed the following opportunities:

- When switching to the year view, the scroll starts at the top and animates down to the active year. 
I think that starting at the active year scroll position would improve the UX. It's the behavior that material.io demonstrates, that I can see on Google Calendar and Gmail (Android).
- When picking a year, like 1900, I think that it would be better to slide the year window around it, so we can choose between 1800 and 2000. It's not ideal to pick a date around 1500, hopefully, the windowing allows moving 200 years at the time is good enough for a marginal problem (very rare).
- Should a tap on the year display the year view? I'm not sure that the dropdown icon surface nor discoverability is good enough.
- It seems that the year chip should be 36px height (OK) and have around borders, what about increasing the border-radius from 16px to 18px?
- The month transition animation on Android is KO. I suspect the view is too heavy. What about removing the ButtonBase from it? I think that it could also greatly help when rendering the month view independently (or when we will need to render 3 months on the same view). Alternatively, we could remove the animation, it's would be fine. Right now, it best benefits iOS users while it harms Android (and harms desktop, but we mentioned that it's a topic for another effort) users.

I haven't noticed anything else, it's so cool to see it making progress :). I haven't looked at the API, but I guess it's too early for that?

dmtrKovalenko commented:
@oliviertassinari 
1. There is a specific prop called `animateYearScrolling` which is by default `false`. It is confusing that it is enabled for the first example, I will remove it. By default it is starting from the year position, but can be usefull for something. 
2. What do you mean? Increse min-max range to include 200 years back and forwards? 
3. Nice questions. I think yes for desktop users click on month should open `MonthView`, click on year open year, but I am not sure about mobile, maybe just tapping on the header should also move to year?
( I have created this functionality on the very first try, but for some reason removed it) 
4. will fix 
5. I propose to create new prop called `reduceAnimations=/(android)/i.test(navigator.userAgent)`. This will actually improve default user experience, and will show animations for desktop and ios. What do you think?

Api changes will appear in the next PR. Here we just moving to new design reference

dmtrKovalenko committed:
Fix luxon tests

dmtrKovalenko committed:
Remove animateYearScrolling from examples

dmtrKovalenko committed:
Do not run docgen on test changes

dmtrKovalenko committed:
Rename data-test-id to data-mui-test and remove test attributes in production

oliviertassinari commented:
1. I didn't know about the `animateYearScrolling` prop. What problem does the prop solve? Do we have an opportunity to reduce the complexity of the component?
2. I was wondering about using the selected year, and not the current year (2019, soon 2020) to generate the list of years people can click on.
3. I have noticed that the date & time picker has an interactive year. I don't have any solution in mind. We might get an issue about this problem in the future. We will see.
5. I like the solution you have proposed. I also think that we should seriously consider giving up on the ButtonBase usage in the day cells. We have to balance between these dimensions:
- can we render 3 months in a reasonable amount of time, on a 4 years old desktop? For instance, I have a 5 years old Macbook Air 🙈. I would have concerns if it takes more than 250ms, for instance, to shift one month.
- can we have instant feedback on keyboard navigation, right now, it feels low. I suspect we won't have any other choice than to render outside React. Try keeping the right arrow down, the UI freeze.
- can we have a smooth animation? Taking into account that no animation might be better than a freezing one.

dmtrKovalenko commented:
1. Yeah I think we can remove this prop. It looks useless.
2. Related to the years. They are generating from the `minDate` to `maxDate`. By default the range is `1900-01-01` to `2100-01-01`. 
3. Yes, it looks like a problem, but still, you can switch to year view from toolbar and also (needs to be implemented) from calendar header 
4. Will try to use `ButtonBase` and check is it reduces performance 

(haha I am seeing that keyboard navigation is stopping after the animation, I suppose the problem is losing focus. Looks like we need a global useKeyDown)

dmtrKovalenko committed:
Remove animateYearScrolling prop

dmtrKovalenko committed:
Improve new props api documentation

oliviertassinari commented:
2. Thanks for the context. I thought the date picker had no limits. What happens if we change the props to remove the min and max date limits? I have tried `<DatePicker minDate={null} maxDate={null}` but it didn't seem to have an impact. https://blueprintjs.com/docs/#datetime/datepicker behaves like the current implementation. (Oh, I see that blueprints went with http://react-day-picker.js.org/, it sounds like its a good signal that we can benchmark with it :)).
What do you think of removing the default limits, and having ±100 years sliding window around the "active" date?

Unrelated, we might have an opportunity to improve a11y. I was wondering about the usage of grid roles for the day grid: https://www.w3.org/TR/wai-aria-practices/#table.

4. I had a closer look at the "calendar" implementation. Do we need the `<div role="presentation">` around the cells? A lighter DOM node could help with performance.

dmtrKovalenko committed:
Fix z-index for in-clock am/pm buttons

dmtrKovalenko committed:
Allow to reduce heavy animations

dmtrKovalenko committed:
Use global keydown event for calendar keyboard navigation

dmtrKovalenko committed:
Restyle highlighting of current dare

dmtrKovalenko commented:
Addressed some feedback:

1) Moving to button base saved 2ms of rendering on my machine (from 10ms to 8ms of rendering dialog with transition group)
2) My concern about the dynamic range by default - will it be obvious for developers? Right now it is clear that `minDate` will not render anything below. Moreover, it will require additional calculations on the rendering step (honestly it should not be a problem) 
3) Yes, we need this wrapper with `role="presentation"`. The main reason is that the day component is overridable, so we need a way to be sure that all event handlers are attached even when the component is overridden. 
4) Keyboard animation was broken by moving to the relative keydown listeners. Right now fixed and will not freeze after switching to the next month. 
5) https://axesslab.com/accessible-datepickers/ I do think that primary way of input day should be a keyboard. I suppose that's why google added an input button even into the mobile datepicker view. I am not sure how users can properly understand the selected date from the calendar day list. I am planning also to significantly improve accessibility by adding proper aria-labels to the date values and properly handle keyboard input. But also it needs to be done after keyboard input from mobile and to be a separate PR

oliviertassinari commented:
1. I have tried the following diff, to benchmark a month switch in dev mode

```diff
diff --git a/lib/src/views/Calendar/Day.tsx b/lib/src/views/Calendar/Day.tsx
index 100cad6..d28a1e8 100644
--- a/lib/src/views/Calendar/Day.tsx
+++ b/lib/src/views/Calendar/Day.tsx
@@ -70,9 +70,9 @@ export const Day: React.FC<DayProps> = ({
   });

   return (
-    <IconButton className={className} tabIndex={hidden || disabled ? -1 : 0} {...other}>
-      <Typography variant="body2" color="inherit" children={children} />
-    </IconButton>
+    <span>
+      {children}
+    </span>
   );
 };
```

**before**
![Capture d’écran 2019-12-22 à 18 56 49](https://user-images.githubusercontent.com/3165635/71325423-d5702600-24ec-11ea-8078-fd1ba3a81d3b.png)

**after**
![Capture d’écran 2019-12-22 à 18 56 41](https://user-images.githubusercontent.com/3165635/71325422-d4d78f80-24ec-11ea-8ebc-b4ceee3f20ba.png)
2. My main concern would be about: if somebody needs to significantly increase the min/max range. How long does it take to render the whole year list (>1s?) and how much to scroll to find the value the user needs?
3. For this problem, we frequently ask for developers to spread the props to the root element, it could be an alternative. I have tried without the wrapping div, it doesn't seem to have a significant impact as we have in 1. It seems that we can ignore performance on the mater. 
4. Nice
5. +1 for keyboard and a11y exploration

dmtrKovalenko commented:
1. I am not seeing such significant performance impact with `ButtonBase` + `<span>` instead of IconButton + Typography. But still removing it in favor of simplicity and fewer dom nodes 
2. Here is a snapshot switching to year selection with rendering a list from 1700 to 2300. Pretty fast, and thanks to animation it is even more smooth. 
![image](https://user-images.githubusercontent.com/16926049/71354865-e4ef7d80-2585-11ea-867c-fa4d52b3468c.png)

Also, I am planning to render 4 years in the column for desktop pickers. For mobile it is too much cause touch targets will be too small. 
> From [material.io](https://material.io/components/pickers/#mobile-pickers): Mobile calendar pickers. They aren’t ideal for selecting dates in the distant past or future that require more navigation, such as entering a birth date or expiration date.
3. Haha, I remember why we need an additional day wrapper :)
```jsx
<DatePicker
  // We are passing already rendered day node right to the override function. Super usefull for adjustign rendering
  renderDay={(day, selectedDate, isInCurrentMonth, dayComponent) => {
     const isSelected = isInCurrentMonth && selectedDays.includes(date.getDate());

     return <Badge badgeContent={isSelected ? '🌚' : undefined}>{dayComponent}</Badge>;
   }}
 />
```

5. Will move this work to separate PR

dmtrKovalenko committed:
Move calendar test to datepicker e2e

dmtrKovalenko committed:
Better styles for Year and Day :focus and :hover

dmtrKovalenko commented:
I think this should be ready to go? Moving to some work in date-io 💪

oliviertassinari reviewed (approved):
One step at the time 👍

dmtrKovalenko committed:
Fix linter
